src/context.rs

```diff
@@ -0,0 +1,146 @@
+use http::{HeaderMap, Method, Uri, Version};
+use http_service::Body;
+use route_recognizer::Params;
+use std::{str::FromStr, sync::Arc};
+
+/// Data associated with a request-response lifecycle.
+///
+/// The `Context` gives endpoints access to basic information about the incoming
+/// request, route parameters, and various ways of accessing the request's body.
+///
+/// Contexts also provide *extensions*, a type map primarily used for low-level
+/// communication between middleware and endpoints.
+pub struct Context<AppData> {
+    app_data: Arc<AppData>,
+    request: http_service::Request,
+    route_params: Params,
+}
+
+impl<AppData> Context<AppData> {
+    pub(crate) fn new(
+        app_data: Arc<AppData>,
+        request: http::Request<Body>,
+        route_params: Params,
+    ) -> Context<AppData> {
+        Context {
+            app_data,
+            request,
+            route_params,
+        }
+    }
+
+    /// Access the request's HTTP method.
+    pub fn method(&self) -> &Method {
+        self.request.method()
+    }
+
+    /// Access the request's full URI method.
+    pub fn uri(&self) -> &Uri {
+        self.request.uri()
+    }
+
+    /// Access the request's HTTP version.
+    pub fn version(&self) -> Version {
+        self.request.version()
+    }
+
+    /// Access the request's headers.
+    pub fn headers(&self) -> &HeaderMap {
+        self.request.headers()
+    }
+
+    /// Access the entire request.
+    pub fn request(&self) -> &http_service::Request {
+        &self.request
+    }
+
+    /// Mutably access the request body
+    pub fn body(&mut self) -> &mut http_service::Body {
```

Comment:
I've generally seen two methods for mutable and immutable access. Any particular reason why we have a single function returning a mutable reference? I ask because I figure the body will only be read more often than not.

Reply:
The reason this one is mutable is because, since the body is a stream, to read it you *must* mutate it.

But more generally you're right that this API design doesn't quite smell right. And in practice, you end up using `take_body` anyway! So I think I'll drop this method for now.